spring-cloud-gcp-pubsub/src/main/java/org/springframework/cloud/gcp/pubsub/reactive/PubSubReactiveFactory.java

```diff
@@ -0,0 +1,168 @@
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.springframework.cloud.gcp.pubsub.reactive;
+
+import java.util.List;
+import java.util.concurrent.TimeUnit;
+
+import com.google.api.gax.rpc.DeadlineExceededException;
+import org.apache.commons.logging.Log;
+import org.apache.commons.logging.LogFactory;
+import reactor.core.Disposable;
+import reactor.core.publisher.Flux;
+import reactor.core.publisher.FluxSink;
+import reactor.core.scheduler.Schedulers;
+
+import org.springframework.cloud.gcp.pubsub.core.subscriber.PubSubSubscriberOperations;
+import org.springframework.cloud.gcp.pubsub.support.AcknowledgeablePubsubMessage;
+import org.springframework.util.Assert;
+
+/**
+ * A factory for procuring {@link Flux} instances backed by GCP Pub/Sub Subscriptions.
+ *
+ * @author Elena Felder
+ *
+ * @since 1.2
+ */
+public final class PubSubReactiveFactory {
+
+	private static final Log LOGGER = LogFactory.getLog(PubSubReactiveFactory.class);
+
+	private PubSubSubscriberOperations subscriberOperations;
+
+	/**
+	 * Instantiate `PubSubReactiveFactory` capable of generating subscription-based streams.
+	 *
+	 * @param subscriberOperations template for interacting with GCP Pub/Sub subscriber operations.
+	 */
+	public PubSubReactiveFactory(PubSubSubscriberOperations subscriberOperations) {
+		Assert.notNull(subscriberOperations, "subscriberOperations cannot be null.");
+		this.subscriberOperations = subscriberOperations;
+	}
+
+	/**
+	 * Create an infinite stream {@link Flux} of {@link AcknowledgeablePubsubMessage} objects.
+	 * <p>The {@link Flux} respects backpressure by using of Pub/Sub Synchronous Pull to retrieve
+	 * batches of up to the requested number of messages until the full demand is fulfilled
+	 * or subscription terminated.
+	 * <p>For unlimited demand, the underlying subscription will be polled at a regular interval,
+	 * requesting up to {@code Integer.MAX_VALUE} messages at each poll.
+	 * <p>For specific demand, as many messages as are available will be returned immediately,
+	 * with remaining demand being fulfilled in the future.
+	 * Pub/Sub timeout will cause a retry with the same demand.
+	 * @param subscriptionName subscription from which to retrieve messages.
+	 * @param pollingPeriod how frequently to poll the source subscription in case of unlimited demand.
+	 * @return infinite stream of {@link AcknowledgeablePubsubMessage} objects.
+	 */
+	public Flux<AcknowledgeablePubsubMessage> createPolledFlux(String subscriptionName, long pollingPeriod) {
+
+		return Flux.create(sink -> {
+			sink.onRequest((numRequested) -> {
+				if (numRequested == Long.MAX_VALUE) {
```

Comment:
I'm still trying to wrap my head around the idea that if someone requests `Long.MAX_VALUE-1`, the behavior is radically different than with `Long.MAX_VALUE`, but I guess `Long.MAX_VALUE` is just has a very special meaning for reactive.

Reply:
Yes, this was the only awkward thing about Reactor; I'd prefer a boolean indicator of unbounded demand. The rest of the framework is perfect elegance, though.